### PR TITLE
Get to one standard for delete operations in WebUI

### LINKED ISSUE #2750: Improve code for WebUI to have single approach for delete() methods (where applicable)
At that moment we have different approaches to delete entity from UI. We can do that using `delete_entity()` method or use whole flow for particular entity deletion procedure. It will be nice to get to common solution and use it in all situations

### PR DESCRIPTION
**Pull request by oshtaier**
Closes #2750

As a bonus:
- Fixed some tests names
- Improved performance for at least on 15 minutes for overall run

```
nosetests tests/foreman/ui/test_activationkey.py -m test_positive_delete_activation_key_with_different_names
.
----------------------------------------------------------------------
Ran 1 test in 266.173s

OK

nosetests tests/foreman/ui/test_activationkey.py -m test_negative_delete_activation_key
.
----------------------------------------------------------------------
Ran 1 test in 50.160s

OK
```
```
nosetests tests/foreman/ui/test_architecture.py -m test_delete_architecture
.
----------------------------------------------------------------------
Ran 1 test in 93.138s

OK
```
```
nosetests tests/foreman/ui/test_computeresource.py -m test_positive_delete_resource
.
----------------------------------------------------------------------
Ran 1 test in 117.956s

OK
```
```
nosetests tests/foreman/ui/test_config_groups.py -m test_delete_positive
.
----------------------------------------------------------------------
Ran 1 test in 82.365s

OK
```
```
nosetests tests/foreman/ui/test_contentviews.py -m test_cv_delete
.
----------------------------------------------------------------------
Ran 1 test in 176.677s

OK
```
```
nosetests tests/foreman/ui/test_discoveryrule.py -m test_positive_delete_discovery_rule
.
----------------------------------------------------------------------
Ran 1 test in 165.295s

OK
```
```
nosetests tests/foreman/ui/test_domain.py -m test_delete_domain
.
----------------------------------------------------------------------
Ran 1 test in 22.444s

OK
```
```
nosetests tests/foreman/ui/test_environment.py -m test_positive_delete_env
.
----------------------------------------------------------------------
Ran 1 test in 42.736s

OK
```
```
nosetests tests/foreman/ui/test_gpgkey.py -m test_positive_delete_for_uploaded_content
.
----------------------------------------------------------------------
Ran 1 test in 167.289s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_positive_delete_for_pasted_content
.
----------------------------------------------------------------------
Ran 1 test in 215.631s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_15
.
----------------------------------------------------------------------
Ran 1 test in 70.726s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_16
.
----------------------------------------------------------------------
Ran 1 test in 81.877s

OK

nosetests tests/foreman/ui/test_gpgkey.py -m test_key_associate_17
.
----------------------------------------------------------------------
Ran 1 test in 88.845s

OK
```
```
nosetests tests/foreman/ui/test_hw_model.py -m test_positive_delete
S
----------------------------------------------------------------------
Ran 1 test in 46.032s

OK (SKIP=1)
```
```
nosetests tests/foreman/ui/test_hostgroup.py -m test_delete_hostgroup
.
----------------------------------------------------------------------
Ran 1 test in 77.444s

OK
```
```
nosetests tests/foreman/ui/test_ldapauthsource.py -m test_positive_delete_ldap_auth_withad
.
----------------------------------------------------------------------
Ran 1 test in 196.968s

OK
```
```
nosetests tests/foreman/ui/test_medium.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 27.539s

OK
```
```
nosetests tests/foreman/ui/test_operatingsys.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 33.183s

OK
```
```
nosetests tests/foreman/ui/test_partitiontable.py -m test_positive_delete
S
----------------------------------------------------------------------
Ran 1 test in 43.458s

OK (SKIP=1)
```
```
nosetests tests/foreman/ui/test_products.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 219.233s

OK
```
```
nosetests tests/foreman/ui/test_puppet_classes.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 79.033s

OK
```
```
nosetests tests/foreman/ui/test_repository.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 288.601s

OK
```
```
nosetests tests/foreman/ui/test_role.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 140.155s

OK
```
```
nosetests tests/foreman/ui/test_subnet.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 94.035s

OK

nosetests tests/foreman/ui/test_subnet.py -m test_negative_delete
.
----------------------------------------------------------------------
Ran 1 test in 25.689s

OK
```
```
nosetests tests/foreman/ui/test_syncplan.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 144.361s

OK
```
```
nosetests tests/foreman/ui/test_template.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 108.324s

OK
```
```
nosetests tests/foreman/ui/test_trend.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 29.338s

OK
```
```
nosetests tests/foreman/ui/test_user.py -m test_positive_delete_user
.S
----------------------------------------------------------------------
Ran 2 tests in 197.374s

OK (SKIP=1)
```
```
nosetests tests/foreman/ui/test_subscription.py -m test_positive_delete
.
----------------------------------------------------------------------
Ran 1 test in 95.796s

OK
```

### CODE REVIEW

**Comment by oshtaier** [comment]
In future we can get to similar approach as we have for search functionality (https://github.com/SatelliteQE/robottelo/issues/2930)

File: robottelo/ui/activationkey.py [hunk 1]
```diff
@@ -131,16 +131,11 @@ def update(self, name, new_name=None, description=None,
...
-
-        if element:
-            element.click()
-            self.wait_for_ajax()
-            self.click(locators['ak.remove'])
-            if really:
-                self.click(common_locators['confirm_remove'])
-            else:
-                self.click(locators['ak.cancel'])
+        self.delete_entity(
```
Comment (elyezer) [inline robottelo/ui/activationkey.py]: Since delete calls delete_entity straight forward is delete_entity needed still?

Comment (elyezer) [inline robottelo/ui/activationkey.py]: Never mind. This answer is yes since the delete is a point to add additional behavior and delete_entity a generic approach.

File: robottelo/ui/contentviews.py [hunk 2]
```diff
@@ -55,18 +55,13 @@ def create(self, name, label=None, description=None, is_composite=False):
         self.wait_for_ajax()
         self.click(common_locators['create'])
 
-    def delete(self, name, really=True):
+    def delete(self, name):
```
Comment (elyezer) [inline robottelo/ui/contentviews.py]: I suppose that the confirmation box is not present here, isn't it?

**Comment by elyezer** [comment]
ACK, clarification asked are not blockers. I'm asking to know about the UI.

File: robottelo/ui/activationkey.py (hunk 1 above)
Comment (oshtaier) [inline robottelo/ui/activationkey.py]: I thought that step is out of scope of current issue as per comment above. We have to remove that dependency, but that will require adding few more variables. I don't mind to work on that after that PR

File: robottelo/ui/contentviews.py (hunk 2 above)
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: We have,but it is completely different and actually is not a message box, but a form inside of UI:
image

Basically, it is not mandatory to push cancel button as it is not a Javascript and you can proceed further and any action (which is not 'Delete' button) will withdraw delete operation. We can use, of course, custom search here and if necessary will do in future, but not now as we don't need to verify that button specifically (or we can do verification of cancel action separately)

File: tests/foreman/ui/test_products.py [hunk 3]
```diff
@@ -217,7 +217,3 @@ def test_remove_prd(self):
                     )
                     self.assertIsNotNone(self.products.search(prd_name))
                     self.products.delete(prd_name)
-                    # Note: refresh is used here because sometimes selenium
-                    # is too fast to check the deleted object and it fails
-                    self.browser.refresh()
```
Comment (sthirugn) [inline tests/foreman/ui/test_products.py]: Are you sure now this is taken care of? I remember the PR in which we specifically added browser refresh.

File: tests/foreman/ui/test_org.py [hunk 4]
```diff
@@ -276,15 +275,9 @@ def test_positive_delete_bz1225588(self):
             # Org cannot be deleted when selected,
             # So switching to Default Org and then deleting.
             session.nav.go_to_select_org('Default Organization')
-            self.org.remove(org_name)
+            self.org.delete(org_name)
             session.nav.go_to_dashboard()
```
Comment (sthirugn) [inline tests/foreman/ui/test_org.py]: Is nav.go_to_dashboard necessary now?

File: tests/foreman/ui/test_org.py [hunk 5]
```diff
@@ -276,15 +275,9 @@ def test_positive_delete_bz1225588(self):
...
-            status = self.org.search(org_name)
-            # Check for at least ten times that org is deleted due #1225588
-            for _ in range(10):
-                status = self.org.search(org_name)
-                if status is None:
-                    break
-            self.assertIsNone(status)
+            self.assertIsNone(self.org.search(org_name))
```
Comment (sthirugn) [inline tests/foreman/ui/test_org.py]: the previous test is not asserting after delete, but this one does.  Can we follow one approach?

File: tests/foreman/ui/test_gpgkey.py [hunk 6]
```diff
@@ -1159,9 +1157,8 @@ def test_key_associate_14(self):
 
     @run_only_on('sat')
     def test_key_associate_15(self):
-        """@test: Create gpg key with valid name and valid gpg key
-        then associate it with empty (no repos) custom
-        product then delete it
+        """@test: Create gpg key with valid name and valid gpg key then
```
Comment (sthirugn) [inline tests/foreman/ui/test_gpgkey.py]: just to let you know - here `@test` is not mandatory.  Testimony considers the first line as the test name now.

File: tests/foreman/ui/test_discoveryrule.py [hunk 7]
```diff
@@ -160,7 +160,7 @@ def test_disable_discovery_rule_1(self):
             self.assertIsNotNone(self.discoveryrules.search(name))
 
     @run_only_on('sat')
-    def test_delete_discovery_rule_1(self):
+    def test_positive_delete_discovery_rule(self):
```
Comment (sthirugn) [inline tests/foreman/ui/test_discoveryrule.py]: I thought we are not repeating the entity name here.  Shouldn't this be just `test_positive_delete`?

File: tests/foreman/ui/test_computeresource.py [hunk 8]
```diff
@@ -231,7 +231,7 @@ def test_update_libvirt_resource_organization(self):
             )
 
     @run_only_on('sat')
-    def test_remove_resource(self):
+    def test_positive_delete_resource(self):
```
Comment (sthirugn) [inline tests/foreman/ui/test_computeresource.py]: `test_positive_delete`?

File: tests/foreman/ui/test_activationkey.py [hunk 9]
```diff
@@ -336,7 +336,6 @@ def test_positive_delete_activation_key_with_different_names(self):
                     )
                     self.assertIsNotNone(self.activationkey.search(name))
                     self.activationkey.delete(name)
-                    self.assertIsNone(self.activationkey.search(name))
```
Comment (sthirugn) [inline tests/foreman/ui/test_activationkey.py]: I assume delete does the assert here correct?

File: robottelo/ui/subscription.py [hunk 10]
```diff
@@ -32,11 +32,10 @@ def upload(self, path, repo_url=None):
         browse_element = self.wait_until_element(locators['subs.file_path'])
         browse_element.send_keys(path)
         self.click(locators['subs.upload'])
-        # Waits till the below locator is visible or until 180 seconds.
         self.wait_until_element(locators['subs.manifest_exists'], 180)
 
     def delete(self):
-        """Uploads Manifest/subscriptions via UI."""
+        """Deletes Manifest/subscriptions via UI."""
```
Comment (sthirugn) [inline robottelo/ui/subscription.py]: wow :+1:

File: tests/foreman/ui/test_products.py (hunk 3 above)
Comment (oshtaier) [inline tests/foreman/ui/test_products.py]: yes, I am sure. Now, it is taken care by:
```
for _ in range(3):
    searched = self.search(name)
    if bool(searched) != really:
        break
    self.browser.refresh()
    if bool(searched) == really:
        raise UIError(
            'Delete functionality works improperly for "{0}" entity'
            .format(name))
```

File: tests/foreman/ui/test_org.py (hunk 4 above)
Comment (oshtaier) [inline tests/foreman/ui/test_org.py]: Will remove. Thanks

File: tests/foreman/ui/test_org.py (hunk 5 above)
Comment (oshtaier) [inline tests/foreman/ui/test_org.py]: Will remove. Thanks

File: tests/foreman/ui/test_activationkey.py (hunk 9 above)
Comment (oshtaier) [inline tests/foreman/ui/test_activationkey.py]: right

File: tests/foreman/ui/test_discoveryrule.py (hunk 7 above)
Comment (oshtaier) [inline tests/foreman/ui/test_discoveryrule.py]: right. I thought to fix that in PR that will fix all names, but you are right. Let me fix that too

File: robottelo/ui/contentviews.py (hunk 2 above)
Comment (sthirugn) [inline robottelo/ui/contentviews.py]: * Is content view only entity which has this kind of a form confirmation?
* self.delete_entity() doesn't seem to handle this situation.  I believe it is better to leave this code as it is.

File: tests/foreman/ui/test_products.py (hunk 3 above)
Comment (sthirugn) [inline tests/foreman/ui/test_products.py]: :+1:

**Comment by sthirugn** [comment]
ACK pending comments.  Great work in this pr.

File: robottelo/ui/contentviews.py (hunk 2 above)
Comment (oshtaier) [inline robottelo/ui/contentviews.py]: It is only one view that has this kind of form confirmation and `self.delete_entity()` handle only positive scenario. Negative scenario is not needed right now and if needed then we will think about proper solution

**Comment by oshtaier** [comment]
Ready for next round of review

File: robottelo/ui/contentviews.py (hunk 2 above)
Comment (sthirugn) [inline robottelo/ui/contentviews.py]: This PR will break the current working logic of content view delete.  I vote to leave content view delete as it is and fix in future PRs if necessary.

File: robottelo/ui/activationkey.py (hunk 1 above)
Comment (elyezer) [inline robottelo/ui/activationkey.py]: Thanks for the clarification.

File: robottelo/ui/contentviews.py (hunk 2 above)
Comment (elyezer) [inline robottelo/ui/contentviews.py]: Either approach works for me. My only concern is to have an issue to track the future improvement.

Comment (oshtaier) [inline robottelo/ui/contentviews.py]: Returned handling of negative scenario as it was previously and sure that we do not need to return all logic as per my previous thoughts

**Comment by elyezer** [comment]
ACK, again. Will wait for @sthirugn to take another look.

**Comment by oshtaier** [comment]
@elyezer Created https://github.com/SatelliteQE/robottelo/issues/3101 to track the progress for further improvement

**Comment by sthirugn** [comment]
ACK

**Comment by lpramuk** [comment]
I dont like this assert in delete_entity() (it is below code touched by this PR - base.py:215)
 ```
            for _ in range(3):
                 searched = self.search(name)
                 if bool(searched) != really:
                     break
                 self.browser.refresh()
             if bool(searched) == really:
                 raise UIError(
                     'Delete functionality works improperly for "{0}" entity'
                     .format(name))
```
Such assert should be in tests side.

**Comment by lpramuk** [comment]
"assert" - I mean check for entity deletion, not really keyword assert

**Comment by lpramuk** [comment]
not all entities like organizations don't get deleted immediatelly...we are "asserting" right away

**Comment by oshtaier** [comment]
@lpramuk 
First, that PR didn't change that logic. (Please, check https://github.com/SatelliteQE/robottelo/pull/2809 (that is actual change) and https://github.com/SatelliteQE/robottelo/pull/3017 (can light up some more details))
I understand your concerns and that you 'don't like' that assert was moved here, but that approach give us at least:
- huge performance boost
- single point for verification logic, so you don't need to duplicate code for lot of entities

>not all entities like organizations don't get deleted immediatelly...we are "asserting" right away
Not sure what you mean, but read more carefully that code:
```
for _ in range(3):
    searched = self.search(name)
    if bool(searched) != really:
        break
```
In case entity is found right away - we will leave loop immediately.

Please, let me know if you need more details

**Comment by sthirugn** [comment]
@lpramuk can you check these lines of code https://github.com/SatelliteQE/robottelo/blob/master/robottelo/ui/base.py#L209#L219 and see if you are still having concerns?  We are waiting until the item gets deleted and timeout is configurable per entity.

**Comment by lpramuk** [comment]
@oshtaier 
I know it wasn't touched but wanted discuss related topic here.
I dont agree with search after deleting entity

* huge performance boost - was brought yes, but it is unrelated to the part of code , lets ignore 
* single point for logic - then it needs to be more clever, many entities specifically organizations dont get deleted immediately....lets discus
tier1 tests are very duplicate,

**Comment by lpramuk** [comment]
```
[tested object] <--- [test]
.delete()                 .assert_deleted()
```
Simply this goes against object model if I assert on [tested object] side within its methods.
We can have common helpers on [test] side for that.

@sthirugn 
I will play with timeouts to see if it can at least workaround orgs deletion issue. Thanks.

**Comment by oshtaier** [comment]
>huge performance boost - was brought yes, but it is unrelated to the part of code , lets ignore - 

hmm. No, we cannot ignore that :)

>single point for logic - then it needs to be more clever, many entities specifically organizations dont get deleted immediately....lets discus

What do you mean? So it will check it three times and do reload of the page. You don't like small delay or that it repeats operation only three times
Specifically organization deleted right away after two attempts (with 1 sec of delay). Provide, specific example. If you mean `test_positive_delete_bz1225588`, so it is a corner case and is out of scope of current task.

**Comment by sthirugn** [comment]
@lpramuk sure and please create a new issue with your concerns/proposals/ideas.

**Comment by oshtaier** [comment]
@lpramuk yeah, please create separate issue and move discussion there

**Comment by oshtaier** [comment]
>Simply this goes against object model if I assert on [tested object] side within its methods.
We can have common helpers on [test] side for that

That was discussed billions of time before. If you have better approach, please propose. I will be happy to help in any possible way

**Comment by lpramuk** [comment]
Yeah we better move to new one. Its closed here. Really :-)
